// custom/elasticSearch.js
function ElasticSearch(isDebug) {
  this._HOST = "http://localhost:9200";
  this._INDEX = "sensefall";
  this._SEPERATOR = "/";
  this._TEMP_TYPE = "<_type>";
  this.request = require('request');

  /* 
  * i.e. http://localhost:9200/sensefall/
  */
  this._DEFAULT_PATH = this._HOST + this._SEPERATOR + this._INDEX + this._SEPERATOR;
  this._isDebug = isDebug;

  this.buildData = function(aX, aY, aZ, gForce, lightLevel, temperature, batteryLevel, dateTime) {
    // '{"x": 0, "y": -1, "z": 30, "g": 1.123, "light": 100, "temp":27, "battery":100, "datetime":1514427497895}'
    this.logging("buildData: creating data object");

    return { "x": aX, 
            "y": aY, 
            "z": aZ, 
            "g": gForce, 
            "light": lightLevel, 
            "temp": temperature, 
            "battery": batteryLevel, 
            "datetime": dateTime };
  }

  this.insertRecord = function(deviceUid, data) {
    var url = this._DEFAULT_PATH + deviceUid;
    this.logging("insertRecord: trying to insert into " + url);
    this.request({
      method: 'POST',
      headers: { 'Content-Type':'application/json' },
      uri: url,
      body: JSON.stringify(data)
    },
    function (error, response, body) {
      if (error) {
        return console.error('insertRecord: failed:', error);
      }
      console.log('insertRecord: success! with:', body);
    });
  }

  this.logging = function(message) {
    if (this._isDebug) {
      console.log("SensePosture: " + message);
    }
  }
}

module.exports = ElasticSearch;
